### ISSUE TITLE
Unhandled Exception: PlatformException(Exception encountered, read, javax.crypto.BadPaddingException: error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT

### ISSUE THREAD

**Issue by JAICHANGPARK** [mogol/flutter_secure_storage#161]
How to fix this problem? 

```
E/flutter (29195): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: PlatformException(Exception encountered, read, javax.crypto.BadPaddingException: error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT
E/flutter (29195): 	at com.android.org.conscrypt.NativeCrypto.EVP_CipherFinal_ex(Native Method)
E/flutter (29195): 	at com.android.org.conscrypt.OpenSSLCipher$EVP_CIPHER.doFinalInternal(OpenSSLCipher.java:570)
E/flutter (29195): 	at com.android.org.conscrypt.OpenSSLCipher.engineDoFinal(OpenSSLCipher.java:351)
E/flutter (29195): 	at javax.crypto.Cipher.doFinal(Cipher.java:1741)
E/flutter (29195): 	at com.it_nomads.fluttersecurestorage.ciphers.StorageCipher18Implementation.decrypt(StorageCipher18Implementation.java:91)
E/flutter (29195): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.decodeRawValue(FlutterSecureStoragePlugin.java:163)
E/flutter (29195): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.read(FlutterSecureStoragePlugin.java:144)
E/flutter (29195): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.access$300(FlutterSecureStoragePlugin.java:29)
E/flutter (29195): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin$MethodRunner.run(FlutterSecureStoragePlugin.java:197)
E/flutter (29195): 	at java.lang.Thread.run(Thread.java:764)
E/flutter (29195): , null)
```

**Comment by bettdouglas** [comment]
I'm also experiencing the same problem 

`message_codecs.dart in StandardMethodCodec.decodeEnvelope at line 572
platform_channel.dart in MethodChannel._invokeMethod at line 161
Called from: <asynchronous suspension>
flutter_secure_storage.dart in FlutterSecureStorage.read at line 30`

PlatformException(Exception encountered, read, javax.crypto.BadPaddingException: error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT
	at com.android.org.conscrypt.NativeCrypto.EVP_CipherFinal_ex(Native Method)
	at com.android.org.conscrypt.OpenSSLCipher$EVP_CIPHER.doFinalInternal(OpenSSLCipher.java:596)
	at com.android.org.conscrypt.OpenSSLCipher.engineDoFinal(OpenSSLCipher.java:363)
	at javax.crypto.Cipher.doFinal(Cipher.java:2055)
	at c.c.a.a.b.b(:8)
	at c.c.a.a.b(:9)
	at c.c.a.a.d(:2)
	at c.c.a.a.onMethodCall(:11)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(:2)
	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(:5)
	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(:2)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(Unknown Source:19)
	at android.os.Looper.loop(Unknown Source:64)
	at android.app.ActivityThread.main(Unknown Source:107)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11)
	at com.android.internal.os.ZygoteInit.main(Unknown Source:274)
)

**Comment by bettdouglas** [comment]
I used this workaround by adding 
`<application
        ...
            android:allowBackup="false"
            android:fullBackupContent="false">`

to AndroidManifest.xml

**Comment by sheychan31** [comment]
> I used this workaround by adding
> `<application ... android:allowBackup="false" android:fullBackupContent="false">`
> 
> to AndroidManifest.xml

@bettdouglas this doesnt seem to work anymore.

**Comment by sheychan31** [comment]
I encounter this too when running the app for the 2nd time.

**Comment by dara1chan** [comment]
I got the same problem.

**Comment by HemilTheRebel** [comment]
I have the same problem. But the bug only triggers some of the time. I always reproduce it
This is my stacktrace: 
```
PlatformException(Exception encountered, read, javax.crypto.BadPaddingException: error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT
I/flutter (30164): 	at com.android.org.conscrypt.NativeCrypto.EVP_CipherFinal_ex(Native Method)
I/flutter (30164): 	at com.android.org.conscrypt.OpenSSLCipher$EVP_CIPHER.doFinalInternal(OpenSSLCipher.java:602)
I/flutter (30164): 	at com.android.org.conscrypt.OpenSSLCipher.engineDoFinal(OpenSSLCipher.java:365)
I/flutter (30164): 	at javax.crypto.Cipher.doFinal(Cipher.java:2055)
I/flutter (30164): 	at com.it_nomads.fluttersecurestorage.ciphers.StorageCipher18Implementation.decrypt(StorageCipher18Implementation.java:91)
I/flutter (30164): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.decodeRawValue(FlutterSecureStoragePlugin.java:163)
I/flutter (30164): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.read(FlutterSecureStoragePlugin.java:144)
I/flutter (30164): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.access$300(FlutterSecureStoragePlugin.java:29)
I/flutter (30164): 	at com.it_nomads.fl
```

**Comment by HemilTheRebel** [comment]
I downgraded to 3.3.3 and the issue disappeared. 
Change pubspec.yaml to `flutter_secure_storage: 3.3.3`

**Comment by rodrigo-martins-ifood** [comment]
I'm having the same problem on 3.3.3

**Comment by iNima** [comment]
I added `<application ... android:allowBackup="false" android:fullBackupContent="false">` with 3.3.4 but the same error is still there.
I had to change back to 3.3.2 and the problem solved.

**Comment by Andrew-Bekhiet** [comment]
I got the same problem with 3.3.3 and 3.3.4, but when downgraded to 3.3.2 it works fine for me

**Comment by sydneyagcaoili** [comment]
Used 3.3.2 then 3.3.1+1 it's still a problem..

**Comment by Andrew-Bekhiet** [comment]
Double check if the version also changed in the pubspec.lock file

**Comment by sandeeppatel1986** [comment]
@Andrew-Bekhiet I have same issue also tried with deleting pubspec.lock file. Any further solution.
Flutter 1.20.2
flutter_secure_storage: ^3.3.2

**Comment by iNima** [comment]
@sandeeppatel1986 Just for double-checking, go to your `.pub_cache/hosted/pub.dartlang.org/` folder and check the version number of flutter_secure_storage. Make sure that the folder name is `flutter_secure_storage-3.3.2`

**Comment by mogol** [comment]
If anyone has consistent failures, could you please provide a steps to reproduce from scratch?

If you change manifest, I believe you have to delete and then reinstall the app with new manifest, simple relaunch shouldn't help.

**Comment by HemilTheRebel** [comment]
@mogol I haven't tried to reproduce this problem due to my schedule currently. I will try to do it today. I haven't changed the manifest in the file since a long time. I had flutter_secure_storage v3.3.3 and it worked. It was when I upgraded to v3.3.4 that I started seeing this error message. I fixed the error by reverting to v3.3.3. My pubspec.yaml has: `flutter_secure_storage: 3.3.3` The fix was removing the caret symbol to select an exact version

**Comment by numen31337** [comment]
Caused a huge failure for us. After thorough investigation and involvement of a number of developers, we were unable to reproduce the problem on more than 10 devices, even though we can see that this happens for almost every user of the application. The only reasonable solution was to completely remove the library.

**Comment by ldemyanenko** [comment]
I'm having this issue on 3.3.4- 3.3.1 versions with Flutter 1.22

**Comment by mrevoir** [comment]
Roughly 1% of my users are having this problem. I haven't found a direct correlation amongst them. I added code at the same time I went from 3.3.3 to 3.3.4 to allow them to report this particular error so it may have been happening before. 

@mogol I'll let you know if I'm able to replicate in a demo project.

For those moving to another library, what did you choose?

**Comment by koskimas** [comment]
This is randomly happening for me too. The payload that causes it for me is a largeish json string that I unfortunately cannot share here.

I appreciate this great library, and thank you for all the work you've done for it. But unfortunately, this kind of bug is a no-go for this kind of a library that we must be able to trust. I realize that you are probably developing this library on your spare time, but I hope you find time to fix this soon.

**Comment by koskimas** [comment]
This also happens with 3.3.3.

**Comment by koskimas** [comment]
Here's a reproduction.

Create a new app using `flutter create`

Replace the content of `main.dart` with this:

```
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Secure storage bug',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Secure storage bug'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final secureStorage1 = FlutterSecureStorage();
  final secureStorage2 = FlutterSecureStorage();

  Future<void> write() async {
    final r = Random();

    for (int i = 0; i < 100; ++i) {
      String payload = 'a';

      secureStorage1.write(key: 'TEST$i', value: payload);
      secureStorage2.write(key: 'TEST$i', value: payload);
    }
  }

  Future<void> read() async {
    for (int i = 0; i < 100; ++i) {
      await secureStorage1.read(key: 'TEST$i');
      await secureStorage1.read(key: 'TEST$i');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              child: Text('write'),
              onPressed: write,
            ),
            RaisedButton(
              child: Text('read'),
              onPressed: read,
            ),
          ],
        ),
      ),
    );
  }
}
```

Click `write` and then `read` (may need to do that couple of times) and you'll get the error.

So I think this is caused by having more than one instance of `FlutterSecureStorage` and using them in parallel. I didn't check if they even need to be used in parallel. Maybe just initializing them in parallel may cause this.

**Comment by koskimas** [comment]
So a temporary fix would be to serialize all calls to this library. Something like this:

```dart
class AsyncMutex {
  Completer<void> _completer;

  Future<void> lock() async {
    while (_completer != null) {
      await _completer.future;
    }

    _completer = Completer<void>();
  }

  void unlock() {
    assert(_completer != null);
    final completer = _completer;
    _completer = null;
    completer.complete();
  }
}

class SecureStorage {
  final _secureStorage = FlutterSecureStorage();
  final _mutex = AsyncMutex();

  Future<void> read({String key}) async {
    try {
      await _mutex.lock();
      return await _secureStorage.read(key: key);
    } finally {
      _mutex.unlock();
    }
  }

  Future<void> write({String key, String value}) async {
    try {
      await _mutex.lock();
      await _secureStorage.write(key: key, value: value);
    } finally {
      _mutex.unlock();
    }
  }
}
```

**Comment by koskimas** [comment]
Looking at the android code, it seems that it's all completely thread-unsafe. [this method](https://github.com/mogol/flutter_secure_storage/blob/develop/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java#L181) is run in a separate thread for each call, and there's no synchronization whatsoever for any method calls except for the `ensureInitStorageCipher` method.

**Comment by Tommimon** [comment]
I'm developing two very similar apps, both using secure storage.

Somehow I encountered this problem in only one of those two apps. I tried all the fixes in this issue but none of those worked for me.
I tried reinstalling secure storage and now both my apps have this issue :(

I really need this package for my project, any alternative?

**Comment by mogol** [comment]
@koskimas thanks for repro case 🎉 , I was looking for it quite a while. Let me try it :+1:

**Comment by koskimas** [comment]
@mogol No problem! Took me a good while too :sweat_smile: 

I think many of the other issues here in github are caused by the thread safety issues too. The cipher classes are not thread safe, but they are called from different threads for each flutter call. Every time two calls run in parallel, things can easily go wrong.

The correct solution, I think, would be to have only one worker thread in android native code, in which all operations are run. That way the code doesn't lock the android main thread, and the code is thread safe. [HandlerThread](https://developer.android.com/reference/android/os/HandlerThread) could be a good solution.

**Comment by koskimas** [comment]
@mogol I'd be happy to provide a PR for fixing the thread safety issues. Are you already working on the fix or should I start?

**Comment by mogol** [comment]
Hey Sami,
I was planning to check on the weekends. So if you can do it early :+1: it’s fine.

**Comment by Sorunome** [comment]
> A temporary workaround would be to serialize all calls to this library and to **only use one `FlutterSecureStorage` instance per app**. Something like this:
> 
> ```dart
> class AsyncMutex {
>   Completer<void> _completer;
> 
>   Future<void> lock() async {
>     while (_completer != null) {
>       await _completer.future;
>     }
> 
>     _completer = Completer<void>();
>   }
> 
>   void unlock() {
>     assert(_completer != null);
>     final completer = _completer;
>     _completer = null;
>     completer.complete();
>   }
> }
> 
> class SecureStorage {
>   static const _secureStorage = FlutterSecureStorage();
>   final _mutex = AsyncMutex();
> 
>   Future<String> read({String key}) async {
>     try {
>       await _mutex.lock();
>       return await _secureStorage.read(key: key);
>     } finally {
>       _mutex.unlock();
>     }
>   }
> 
>   Future<void> write({String key, String value}) async {
>     try {
>       await _mutex.lock();
>       await _secureStorage.write(key: key, value: value);
>     } finally {
>       _mutex.unlock();
>     }
>   }
> }
> ```

If you make the mutex global you can also use multiple instances, in case it is non-trivial to adapt your app to have a single instance for now and you need a quick fix.

**Comment by koskimas** [comment]
@Sorunome I updated the example :+1:

**Comment by koskimas** [comment]
Here's a fix https://github.com/mogol/flutter_secure_storage/pull/165

**Comment by Philipus13** [comment]
> Here's a fix #165

Please merge this PR immediately

**Comment by mogol** [comment]
Hey guys,
Please try 3.3.5, it has been published. Kudos to @koskimas

**Comment by XinyueZ** [comment]
The issue causes a great user lost, sorry, had to remove the library and replace it with other stable ones.

**Comment by JAICHANGPARK** [comment]
@koskimas  @mogol and Thanks All

**Comment by vdmitrovskiy-lohika** [comment]
Hi! 
With 3.3.5 got this error. 
I got this error only when deploying app to a phone from Google play. On simulator or uploading apk directly, this error doesn't fire.

**Comment by Kacppian** [comment]
I'm having an issue with the newer version as well. We're using 3.3.5 and it happens randomly.

**Comment by Rebloom** [comment]
I wonder if the switch isStrongBox support caused this,
It happened when use Google Device with TitanM chip.
Wanna know the bug device models.

**Comment by Rebloom** [comment]
Here are the associate links:
https://stackoverflow.com/questions/59552385/aes-key-with-setisstrongboxbacked-failed-decryption
https://stackoverflow.com/questions/62639755/android-strongbox-backed-keys-cannot-be-used-for-rsa-encryption
https://alexbakker.me/post/mysterious-google-titan-m-bug-cve-2019-9465.html

**Comment by julianzur** [comment]
@Rebloom: I just ran into the issue this weekend again with 3.3.5 on a Pixel 4a with latest security updates from Nov 2020. 
According to your sources the TitanM problem ["was fixed in the Android security update of December 2019 for the Pixel 3 and newer."](https://stackoverflow.com/a/60465658). Do you have any news if the problem could've re-appeared?

**Comment by Rebloom** [comment]
> @Rebloom: I just ran into the issue this weekend again with 3.3.5 on a Pixel 4a with latest security updates from Nov 2020.
> According to your sources, the TitanM problem ["was fixed in the Android security update of December 2019 for the Pixel 3 and newer."](https://stackoverflow.com/a/60465658). Do you have any news if the problem could've re-appeared?

We meet this issue randomly, but both 2 devices are google 3a+ device, I wonder there is still unfixed problem with google secure chip, cause other Samsung/Xiao Mi devices would never occurs this.Especially the sync code about google soft secure zone to physical chip like TitanM, too hard to catch. You can check up Google Android isStrongBox related code for reference. I am hoping people that all ran into this issue could reveal the device Model so we can check up some common meet.

**Comment by Rebloom** [comment]
> @Rebloom: I just ran into the issue this weekend again with 3.3.5 on a Pixel 4a with latest security updates from Nov 2020.
> According to your sources, the TitanM problem ["was fixed in the Android security update of December 2019 for the Pixel 3 and newer."](https://stackoverflow.com/a/60465658). Do you have any news if the problem could've re-appeared?

TitanM chip could not be upgrade until a certain authentication, I think If remove isStrongBox feature would solve this question,We can happily use This secure_storage,rather than without Android's trust zone.Just in my opinion. Hope this issue fixed soon.

**Comment by Soda-Flavour** [comment]
After updating to 3.3.5 for this problem, 2 of the 3 devices that failed have been fixed, but 1 still has errors.
Therefore, version 3.3.5 still has an error.

**Comment by koskimas** [comment]
@Soda-Flavour Did you delete the app and reinstall it? Because the problem is that the old version corrupted the data. No amount of updates will fix the data if it's corrupted. Does 3.3.5 work if you start from a clean slate with no data in the storage?

**Comment by koskimas** [comment]
People that are getting this error in production but not in development: Are you absolutely sure you have the correct version in the lock file and the CI, or whatever you use for building the production package, actually picks up the correct version?

**Comment by Soda-Flavour** [comment]
> People that are getting this error in production but not in development: Are you absolutely sure you have the correct version in the lock file and the CI, or whatever you use for building the production package, actually picks up the correct version?

1. Situation
The production app has identified the error code.
So we updated the version to 3.3.5 and confirmed it in the .lock file.
The device that has an error is LG v50.
After the update, we deleted the app, installed it, and restarted the device.

2.doctor
[✓] Flutter (Channel table, 1.20.2 on Mac OS X 10.15.5 19F101, locale ko-KR)
• Flutter version 1.20.2 at /Users/amoskim/flutter
• Framework revision bbfbf1770c (4 months ago), 2020-08-13 08:33:09 -0700
• Engine revision 9d5b21729f
• Dart version 2.9.1


Do I have to check or do I miss anything?

**Comment by narritt** [comment]
Got the same problem with Samsung S10 (Android 10).
[√] Flutter (Channel stable, 1.22.5, on Microsoft Windows [Version 10.0.19042.631], locale ru-RU)
    • Flutter version 1.22.5 at C:\flutter
    • Framework revision 7891006299 (6 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

pubspeck.yaml and pubspec.lock also have version 3.3.5. I added this package to my project on version 3.3.5 so I think it contains only the last plugin version

**Comment by Nailer** [comment]
Getting this error now on an app in production which makes the whole app freeze on launch for a portion of our users.
Works on Pixel 1 (Android 10), Pixel 2 (Android 11), but fails on my Pixel 5(Android 11).

**Comment by gabdsg** [comment]
I still have this problem too

**Comment by julianzur** [comment]
We fixed it by forking the project and removing the strongbox feature. So far we don't see any problems anymore.

**Comment by chitgoks** [comment]
This was solved with

android:allowBackup="false"
android:fullBackupContent="false"

in AndroidManifest.xml

**Comment by DzmitryBarysevichLeverX** [comment]
I also have this issue.
But I find out interesting thing.
It has worked when I had a few libraries and small project.
But after I have installed firebase I have started to catch this error.

My libraries
`
  cupertino_icons: ^1.0.0
  provider: ^4.3.3
  http: ^0.12.2
  flutter_secure_storage: ^3.3.5
  modal_progress_hud: ^0.1.3
  infinite_scroll_pagination: ^2.3.0
  intl: ^0.16.1
  dropdown_search: ^0.4.8
  firebase_core: ^0.7.0
  firebase_analytics: ^7.0.1
  firebase_crashlytics: ^0.4.0+1
`

**Comment by DzmitryBarysevichLeverX** [comment]
I also have this issue.
But I find out interesting thing.
It has worked when I had a few libraries and small project.
But after I have installed firebase I have started to catch this error.

My libraries:
```
  cupertino_icons: ^1.0.0
  provider: ^4.3.3
  http: ^0.12.2
  flutter_secure_storage: ^3.3.5
  modal_progress_hud: ^0.1.3
  infinite_scroll_pagination: ^2.3.0
  intl: ^0.16.1
  dropdown_search: ^0.4.8
  firebase_core: ^0.7.0
  firebase_analytics: ^7.0.1
  firebase_crashlytics: ^0.4.0+1
```

**Comment by nikita-filimonau** [comment]
I also have this issue.
But I find out interesting thing.
It has worked when I had a few libraries and small project.
But after I have installed firebase I have started to catch this error.

My libraries:
```
  cupertino_icons: ^1.0.0
  provider: ^4.3.3
  http: ^0.12.2
  flutter_secure_storage: ^3.3.5
  modal_progress_hud: ^0.1.3
  infinite_scroll_pagination: ^2.3.0
  intl: ^0.16.1
  dropdown_search: ^0.4.8
  firebase_core: ^0.7.0
  firebase_analytics: ^7.0.1
  firebase_crashlytics: ^0.4.0+1
```

**Comment by RANSoftRA** [comment]
I am experiencing the same issue with a production app (also flutter_secure_storage 3.3.5).

**Comment by mahdi-ninja** [comment]
Seems like the issue is fixed on older Pixel 3 phones but we're still having the same issue on Pixel 4a. Tried every suggestion but didn't work for us and ended up going back to insecure storage (i.e. shared_preferences)

**Comment by RANSoftRA** [comment]
I have created a new issue #210, since this problem still seems to be active.

**Comment by vanlooverenkoen** [comment]
I am having this issue on a huawei mate 20 lite

**Comment by jasaw** [comment]
@mahdi-ninja I have the opposite observation. It's failing on my Pixel 3, but works fine on my Pixel 4a. I'm running flutter_secure_storage version 3.3.5, both clean install, production build of app (not debug).

**Comment by postpasspost** [comment]
This error happens only in production in the play store. Huawei android 10. Tried this:

> If you need `fullBackupContent="yes"`, you can [disable backup](https://developer.android.com/guide/topics/data/autobackup#IncludingFiles) of prefs used by the plugin.
> 
> ```
> <?xml version="1.0" encoding="utf-8"?>
> <full-backup-content>
>     <exclude domain="sharedpref" path="FlutterSecureStorage"/>
> </full-backup-content>
> ```

But I still get the error. Using `flutter_secure_storage: ^4.1.0`

**Comment by komritza** [comment]
I am using version 3.3.5 in the production and since few days ago I started getting sentry reports with following error log:
``` 
PlatformException
PlatformException(Exception encountered, read, javax.crypto.BadPaddingException: error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT
	at com.android.org.conscrypt.NativeCrypto.EVP_CipherFinal_ex(Native Method)
	at com.android.org.conscrypt.OpenSSLEvpCipher.doFinalInternal(OpenSSLEvpCipher.java:152)
	at com.android.org.conscrypt.OpenSSLCipher.engineDoFinal(OpenSSLCipher.java:374)
	at javax.crypto.Cipher.doFinal(Cipher.java:2055)
	at c.c.a.b.b.b(Unknown Source:32)
	at c.c.a.a.b(Unknown Source:11)
	at c.c.a.a.d(Unknown Source:7)
	at c.c.a.a.a(Unknown Source:0)
	at c.c.a.a$b.run(Unknown Source:134)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:246)
	at android.os.HandlerThread.run(HandlerThread.java:67)
, null)
```

I tried to reproduce the bug, and I manage to do it. As mentioned above, I have uninstall app from my phone(samsuns s10e), and downloaded the app directly from the PlayStore and app got frozen on the splash screen and I got an instant sentry report. App stopped completely working...
Then I have installed the app from VS code and it started working again, but I got this log in my terminal:
``` 
D/FlutterSecureStoragePl( 7250): Initializing StorageCipher
I/fluttersecurestorage( 7250): Creating keys!
I/fluttersecurestorage( 7250): Initializing
I/fluttersecurestorage( 7250): Generating key pair
E/KeyStore( 7250): generateKeyInternal failed on request -68
D/CertificatePolicyCache( 7250): Creating new instance of CertificatePolicyCache myUid: 10444 callingUid: 10444
D/CertificatePolicyCache( 7250): readVariables type: null userId: 0
D/FlutterSecureStoragePl( 7250): StorageCipher initialization complete
```
I don't know if this log is helpful, but this is actually a really huge bug.

**Comment by dluanoliveira1** [comment]
Hey guys, the problem might be setting a key as empty String, at least for me that was it.

**Comment by loonix** [comment]
I think @dluanoliveira1 got it, for example I am using this plugin and at the start of the app I am reading a value that might be stored. But when I first start the app (after 1st install) there are no keys registered yet, so it gives back that BAD DECRYPT error when it tries to read a key that does not exist.

**Comment by DineshKachhot** [comment]
I am still facing the same issue in release build, I am using flutter_secure_storage: ^4.2.0

**Comment by appu-sm** [comment]
I'm still facing the issue in v 4.2.1 and none of the above solutions work

**Comment by matthewrice345** [comment]
Just ran into this issue with a Google Pixel 6 Pro tried with:
`flutter_secure_storage: ^5.0.0-beta.5` and
`flutter_secure_storage: ^4.2.0`

**Comment by kjawadDeveloper2** [comment]
I am also facing this error.

**Comment by jesusmartinezd** [comment]
Same error here

**Comment by okan-oz** [comment]
Same issue with flutter_secure_storage: 4.2.0

**Comment by Nazarii77** [comment]
Still same issue with flutter_secure_storage: 5.0.2  (but happens only on some real devices, never on emulator )

**Comment by loonix** [comment]
Have you checked if you are reading a value that does not exist?

**Comment by AG-IT-Solutions** [comment]
Having the same Problem with 5.0.2 and i am, too, trying to read a key that doesn't exist yet. Is there a way to check first if the key exists or not? If not, i guess i'll have to create it first with a dummy value and check if it's the dummy value in there or the one i need.

**Comment by CrownKingClown** [comment]
This line caused the error to me 
`String value = await _storage?.read(key: "username");` --> version 4.2.0

**Comment by FreedomLiberty1776** [comment]
I had the same problem with version ^4.2.1 and ^5.0.2 on Samsung J737A which didn't use to happen for a while until I upgraded all my packages. I realized the problem was caused by reading keys that did not yet exist. I solved it by use a simply try catch, and returned null on error, (I think null is the default return value when a key is not found anyways).

static Future getToken() async {
    try {
      return await _storage.read(key: 'token');
    } catch (e) {
      return null;
    }
  }

**Comment by jfahrenkrug** [comment]
If you use version `5.0.2` or newer, make sure you use this option when creating the instance:

```
FlutterSecureStorage(aOptions: AndroidOptions(
    encryptedSharedPreferences: true,
));
```

This causes the package to use Android's built-in EncryptedSharedPreferences. The crash only seems to occur when not using this setting, because then the package uses a different way of encrypting and decrypting the prefs. You can see that by reading the source code here: https://github.com/mogol/flutter_secure_storage/blob/26efe91a75228ad8c8626d6eea18f7f3cb21bdd9/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java#L101

Also see https://github.com/mogol/flutter_secure_storage/pull/328#issuecomment-988577971

**Comment by ubaidillahSriyudi** [comment]
> I used this workaround by adding `<application ... android:allowBackup="false" android:fullBackupContent="false">`
> 
> to AndroidManifest.xml

Thanks it works for me

**Comment by sajanIocod** [comment]
I am still experiencing this issue. Any ideas on how to resolve it?
